### ISSUE TITLE
implement pan action

### ISSUE THREAD

**Issue by filod** [rrag/react-stockcharts#51]
As said [here](https://github.com/rrag/react-stockcharts/blob/next/docs/md/CLICK-CALLBACK-INTERACTIVE-INDICATOR.md), just don't know when will this be implemented or where to track.

**Comment by rrag** [comment]
RE: implementing pan, zoom events to the `ClickHandler`

What is the use case for triggering a click event when panstart/pan/panend happens?

**Comment by filod** [comment]
such as ' drag to load more data '

**Comment by iNikNik** [comment]
@rrag any tips how to implement 'load more data' when user drags to the end of current data? I believe you cannot access current visible data inside onPan handler

**Comment by rrag** [comment]
It will need changes to [here](https://github.com/rrag/react-stockcharts/blob/932c06cfe100f35f325b335cb97c7bf9cd704920/src/lib/scale/evaluator.js#L116-L118)

that `if` block will get executed when pan is no longer possible. You might have to pass in a callback and invoke it from that `if` block.

It is tricky because say you pass a callback and invoke it from there, you would set the state of the owner react component, but the `domain` [function](https://github.com/rrag/react-stockcharts/blob/932c06cfe100f35f325b335cb97c7bf9cd704920/src/lib/scale/evaluator.js#L46) would not have completed, so it could lead to an inconsistent state of `ChartCanvas` and `EventHandler`

I am currently working on the [mouse hover](http://rrag.github.io/react-stockcharts-next/documentation.html#/hover_tooltip) and after that planning to work on algorithmic automated Buy/sell indicators.

Let me think through this and I will post back here. Do you have a different suggestion?

**Comment by iNikNik** [comment]
@rrag thanks for your feedback. I'll try to figure it out. Hope this'll work)

**Comment by iNikNik** [comment]
@rrag it seems this `if block` have never executed. But there is simple solution for me - I've inserted code below before [this block](https://github.com/rrag/react-stockcharts/blob/932c06cfe100f35f325b335cb97c7bf9cd704920/src/lib/scale/evaluator.js#L51)
```javascript
if (left < 0) {
   loadMoreData()
}
```

But, unfortunately I need to change EventHandle - when I'm adding new data to the beginning it logs `Trivial change` and dont want to update visible data.

**Comment by rrag** [comment]
I am sorry, I need to test this a bit to answer.

I like this idea of drag for more data, I will add it to the scope of `0.5`

**Comment by rrag** [comment]
Are you trying to replicate the behavior of tradingview.com? where after initial chart load, when you pan left, the new points are not shown till you release the mouse and complete the pan?

If so, then I think you will have to make changes to the `handlePanEnd` to invoke a callback after the `setState`

**Comment by iNikNik** [comment]
> where after initial chart load, when you pan left, the new points are not shown till you release the mouse and complete the pan?

Yes, it's exactly what I want. 

> If so, then I think you will have to make changes to the handlePanEnd to invoke a callback after the setState

It seems like I can compare `plotData[0].date.getTime() === fullData[0].getTime()` to detect when user have paned to beginning. Thanks again :+1: 

But I still need to modify [this block](https://github.com/rrag/react-stockcharts/blob/master/src/lib/EventHandler.js#L141-L176), because overwise it says `Trivial change` after data have pasted to the beginning.

It seems I need to use current end of interval, but modify start of showing interval.  Also [this condition](https://github.com/rrag/react-stockcharts/blob/master/src/lib/EventHandler.js#L141-L142) should looks like:
```javascript
} else if (dataAltered
              && (
                isLastItemVisible(this.props.fullData, this.state.plotData)
                || isFirstItemVisible(this.props.fullData, this.state.plotData)
)) {
```

or I should create another if block `for isFirstItemVisible` .. I'll test it now

**Comment by rrag** [comment]
how about changing [this block](https://github.com/rrag/react-stockcharts/blob/master/src/lib/EventHandler.js#L177-L185) to set `plotData` to state.

you have access to the `xScale.domain()` and you could get the new `plotData` from the `nextProps.fullData`

**Comment by rrag** [comment]
also in the `handlePanEnd` you should add your code in [this block](https://github.com/rrag/react-stockcharts/blob/master/src/lib/EventHandler.js#L616-L618)

from

```js
		}, () => {
			if (isDefined(callbackList)) callbackList.forEach(callback => callback());
		});
```
to
```js
		}, () => {
			if (isDefined(callbackList)) callbackList.forEach(callback => callback());

			/* get displayXAccessor from context, or change `ChartCanvas` to pass it as props to `EventHandler`
			if (xScale.domain()[0] < xAccessor(first(this.state.plotData)) 
			           && displayXAccessor(first(this.state.plotData)) === displayXAccessor(first(this.props.fullData)) { 
			    loadMoreData()
*/
		});
```

**Comment by rrag** [comment]
@iNikNik how did it go?

**Comment by iNikNik** [comment]
@rrag hello! I've finally did it. Today or tomorrow I'll commit this changes to my fork and I hope you'll make a little review ;)

**Comment by itsjimbo** [comment]
First thank you so much for this library, very awesome.
Is there a way, in keeping with your examples of ClickCallback or Brush to have ChartPanEndCallback?  Purpose is to also load data as above.
  
Implementation may something like
```
<ChartCanvas ..
      <Chart ...
         <ChartPanEndCallback enabled={true}  onPanEnd={ e => { console.log(`load more data=`, e); } }/>
```
I am not sure where the most appropriate place to have a "onPanEnd" callback, I have been playing with adding onPanEnd event to the EventCapture render method, then in a similar fashion creating a callback on a ChartPanEndCallback component.  Is this the best way to proceed?

Thanks 
-Jim

**Comment by rrag** [comment]
adding to `EventCapture` is how I would do it.

**Comment by iNikNik** [comment]
@rrag I'm so sorry ... i've just disappeared ... having a lot of business. 

So, I've added this condition to handle when the data has been added to the beginning. It's about 'loading more data' on pan end. https://github.com/iNikNik/react-stockcharts/blob/feature-infinite-loading/src/lib/EventHandler.js#L149-L150

Main goal is - we should increase indexes.

**Comment by rrag** [comment]
I like the name infinite-loading :)

I agree with the problem, the index is pre calculated so adding data to the beginning screws it up.

**Comment by iNikNik** [comment]
@rrag hello again!) Here is working implementation:
![react-stockcharts-infinite-loading](https://cloud.githubusercontent.com/assets/7679749/16498459/a2bb245c-3f04-11e6-81d1-0c3a14471aa8.gif)

I've added [new prop callback](https://github.com/iNikNik/react-stockcharts/blob/feature-infinite-loading/docs/lib/charts/CandleStickChartWithInfiniteLoading.jsx#L84) `onLoadMoreData` to EventCapture:
```javascript
 <EventCapture mouseMove={true} zoom={true} pan={true} onLoadMoreData={onLoadMoreData} />
```

Also, inside EventHandler  componentWillRecieveProps now looks like:
```javascript
if(reset) {
  // ...
} else if(dataAltered && isDataInsertedToTheBeginning) {
  // increase indexes
} else if (dataAltered && isLastItemVisible) {
 // no changes
}
```

I want you to take a look at [isDataInsertedToTheBeginning](https://github.com/iNikNik/react-stockcharts/blob/feature-infinite-loading/src/lib/EventHandler.js#L42-L47) function. Because we should recognize - is the first item changed? But we can't you xAccessor (it always returns 0), also we can't use date accessor ... or shallowEqual (because if we're using indicators like macd - our data point will have `macd` property and it is object. And yeah, this object is diferent because recalculating props). Only way is - just use deepEqual. What do you think about it?

You can checkout my [fork](https://github.com/iNikNik/react-stockcharts/tree/feature-infinite-loading) and tale a look at http://localhost:8080/documentation.html#/infinity_loading - live example.

**Comment by rrag** [comment]
Thanks @iNikNik I will look at this. I like the results, however the `next` version has done away with `fullData`. I like this feature, and will take inspiration from what you have done to implement this here.

Unfortunately it will not be a simple merge as `next` has diverged so much from master now.

**Comment by iNikNik** [comment]
@rrag oh, I understand. I also can investigate `next` branch - because react-stockcharts is a big part of the project I'm currently working on.  And I want to keep my fork as close to main lib as possible.

**Comment by rrag** [comment]
@iNikNik @filod 

I have been trying to implement this in 0.6 now, there are too many edge cases and the program complexity increases, it makes me wonder if this is really worth it. What are we saving by doing a pan to load more of historical data? I can understand pan to load more for future data, but it beats me why this is useful for historical.

Some problems I am seeing so far.

1. If a chart has a few indicators on it, by providing just the data to render on screen, most of the indicators are not warmed up. Meaning if you already have a EMA 50, the first 50 candles will not have any values, as data prior is not provided in the input prop. To solve this we have to calculate the indicators serverside, and send the `data[]` which has the values of different indicators. Though it is  possible to pre calculate the `data []` serverside, the code still needs the `calculator` for other purposes, so we need to figure out a work around for this

1. For advanced charts like renko/p&f the exact number of rows to be downloaded cannot be determined by the pan end action, so this means, the P&F, Renko calculation has to be done server side too

1. In the event of using a discontinuous scale (it is needed for Renko/p&f) the index which contains the converted linear scale, also has to be calculated on server side, and sent across to the browser so the same scales are used

While all of this can certainly be done, it just makes it technically challenging and makes me wonder if this is all worth it. So I want to hear why you guys think this is an important feature and why it is better than just downloading all historical data.

**Comment by iNikNik** [comment]
@rrag my point is - just pass loaded data as input prop:
1) pan to the end of history data
2) fire onLoadMore event
3) fetch data from server
4) rerender chart with new data prop (nextData = [ ...loadedData, ...data ]

It looks like:

[InfiniteLoadingPage](https://github.com/iNikNik/react-stockcharts/blob/feature-infinite-loading/docs/lib/page/InfiniteLoadingPage.jsx#L33)
```javascript
<Section colSpan={2}>
  <TypeChooser>
    {(type) => <CandleStickChartWithInfiniteLoading data={this.props.someData} type={type} />}
  </TypeChooser>
</Section>
```

[CandleStickChartWithInfiniteLoading](https://github.com/iNikNik/react-stockcharts/blob/feature-infinite-loading/docs/lib/charts/CandleStickChartWithInfiniteLoading.jsx#L84)
```javascript
<EventCapture mouseMove={true} zoom={true} pan={true} onLoadMoreData={onLoadMoreData} />
// ...


```

**Comment by rrag** [comment]
Sure, I see how you have done this, but the problem is still there. You are passing only the last [180](https://github.com/iNikNik/react-stockcharts/blob/feature-infinite-loading/docs/lib/charts/infiniteLoadingWrapper.jsx#L11) elements, so say there is a MACD indicator on the chart, it would not show the values for the first 26 elements as it would be warming up. To avoid this you would have to pre calculate the indicators and then slice the data.

Also to make this work perfectly, you need to download only x elements, where x is decided by the `EventCapture` and deciding the value of x for Renko, P& F is impossible for `EventCapture`. This is what I mean by many edge cases.

While I do like a challenge, and I am quite inclined to work on this, I want to understand the real business need and value of this feature.

In my opinion this feature is just for kicks and challenging how flexible the library is.

**Comment by rsklyar** [comment]
can't wait when this functional available in 0.6 repo.

there are lot of value of this future - loading history from intraday market data (tick by tick).

**Comment by iNikNik** [comment]
@rrag 
> it would not show the values for the first 26 elements as it would be warming up

yep, I see. But now (without this feature), MACD is not visible for the first 26 elements. So, there would be no changes at this point.

> Also to make this work perfectly, you need to download only x elements

Probably we should keep it in the userland. For example - user can select date range for chart:
![image](https://cloud.githubusercontent.com/assets/7679749/17663392/dc3b752e-62f4-11e6-9c05-4e881afe9ba0.png)
If 1d selected - I want to load 1 day of historical data every time. So, it hard to recognize inside stockcharts what I exactly want to load.

**Comment by rrag** [comment]
> yep, I see. But now (without this feature), MACD is not visible for the first 26 elements. So, there would be no changes at this point.

No, you misunderstood, as of now, the calculation of MACD happens on the chart with the entire data, so the first 26 candles do not have MACD data, but while implementing this, say I have 7000 rows of data but I send only 200 datapoints to the chart, the first 26 of this 200 do not have MACD, while in the current situation, we will have the first 26 of the 7000 will not have MACD, which is ok because we do not even have data before 7000. So my point is that the data will have to be precalculated with all the indicators.

This will not be straightforward to implement without a server, so I am mocking up some `setTimeout`s to emulate server which does the pre calculation.

**Comment by rrag** [comment]
work in progress version [here](http://rrag.github.io/react-stockcharts-next/documentation.html#/load_more_data_on_pan)

loading data on zoom out, when you click and drag xAxis to zoom is not implemented yet. 

Might still have a few rough edges, but over all it seems to work.

The example 

1. pre calculates the discontinuous scale and the calculators
1. when you pan to left, it downloads the exact number of rows needed to fill the gap.

have not tested for Renko/P&F yet, will do that tomorrow later.

Appreceate feedback please

**Comment by rrag** [comment]
@iNikNik feedback please

**Comment by iNikNik** [comment]
@rrag I have some troubles to run "next" version:
Sometimes modernizr don't want to start (window is not defined exception), after some magick I've finally started "watch", but debugger (chrome) is not working. Maybe I missed something?

**Comment by iNikNik** [comment]
@rrag Also, I've finally realized what you need to manually pass indicators stuff with data... never used indicator before and I thought it calculates inside chart O_o. Sorry for stupid questions then)

**Comment by rrag** [comment]
I have been the past few days been working on creating an open source version of the demo, which has a server too, and these calculations can happen on the server. While I agree with you have to do more work here, I think you get biased by seeing all that code in the same component, basically all [this](https://github.com/rrag/react-stockcharts/blob/next/docs/lib/charts/CandleStickChartPanToLoadMore.jsx#L91-L106) will move to the server, and it would send 

1. the new data, 
2. the calculated indicators
3. the index for updating the scale

once all these 3 are available the chart should be updated.

So if that code is moved to server, the client will simply have to get a new `xScale` with the index from 3. above, and merge 1, 2 and 3 from above into a new `data []`. This way, say if you add a new indicator to an existing chart, that information is sent to server and the server calculates the indicator for just the rows needed, and send just that information back. Now client has to merge that with `data` and render again.

Though I had my reservations earlier, I am starting to like this server calculated indicators, it could bring some efficiencies, like

1. All the calculators can be removed from the bundle, thus reducing the size of the js file
1. These calculators does not have to be in js, they can be in R or Python or whatever I could potentially use any of the existing third party packages in any language to calculate all these indicators
1. client cpu cycles are not wasted in calculating the data

cc @debsush since you wanted to calculate the indicators using R
cc @itsjimbo since you wanted to calculate only the new indicators, and just for the newly added data points

**Comment by debsush** [comment]
Superlikes!!!!!

**Comment by rrag** [comment]
looking at https://github.com/multicharts/tvbasicstudies/tree/master/src it appears to be server generated have to debug through the xhr and websocket traffic to be sure though

**Comment by debsush** [comment]
Wonder if the transfer of data and repeated client-server-client cycles slow down the app.

**Comment by rsklyar** [comment]
data calculation on server side on traders software - bad idea.

**Comment by rrag** [comment]
merged to master and `0.6.0-beta.2` published to npm
